### PR TITLE
Holiday stop for Guardian Weekly with N-for-N scenario

### PR DESCRIPTION
**Pull request by mario-galic**
Holiday processor requires `chargedThroughDate` to be populated, that is, subscription has been invoiced, in order to calculate when to apply holiday credit (next billing period). Regular Guardian Weekly subscription is always invoiced and hence always has `chargedThroughDate`. However when the reader is still on N-for-N  Guardian Weekly introductory plan (for example, GW Oct 18 - Six for Six - Domestic), then `chargedThroughDate` of corresponding regular Guardian Weekly is not populated, because it will not be invoiced until N-for-N is finished.

- All the changes are in `CurrentGuardianWeeklySubscription.scala`
- No changes were made to`CurrentGuardianWeeklySubscription` model which isolates the rest of the client code from this change
- There is no change to processing of regular GW, that is, new code paths are simply a lazy fallback:

      (maybeRegularGw orElse maybeGwWithIntro)

- `CurrentGuardianWeeklySubscription.invoicedPeriod` is now set directly from `chargedThroughDate` or predicted of N-for-N RPC's `chargedThroughDate`:

      billingPeriod match {
        case "Quarter" => Some(gwNForNChargedThroughDate.plusMonths(3))
        case "Annual" => Some(gwNForNChargedThroughDate.plusYears(1))

- Predicted calculation is tested by `PredictedInvoicedPeriodSpec.scala`

### CODE REVIEW

**Comment by mario-galic** [comment]
[Fix holiday stop for GW N-for-N](https://trello.com/c/EJAhAax6/355-fix-holiday-stop-for-gw-n-for-n)

File: lib/holiday-stops/src/main/scala/com/gu/holiday_stops/Config.scala [hunk 1]
```diff
@@ -113,6 +114,45 @@ object Config {
...
+    "2c92c0f95a24621b015a3d6ce9e32057", // "name":"Guardian Weekly 6 Issues"}
+    "2c92c0f95a246220015a3d680fe1680d", // "name":"Guardian Weekly 6 Issues"}
+    "2c92c0f95a4b489a015a5be867ae7445", // "name":"Guardian Weekly 12 Issues"}
+
+    "2c92c0f8660fb5dd016610858eb90658", // "name":"GW Oct 18 - Six for Six - Domestic"}
+    "2c92c0f9660fc4c70166109dfd08092c", // "name":"GW Oct 18 - Six for Six - ROW"}
+
+  )
+
+  val gwNforNProductRatePlanIdsPROD = List(
```
Comment (kelvin-chappell) [inline lib/holiday-stops/src/main/scala/com/gu/holiday_stops/Config.scala]: These `gwNforN...` vals could do with some explanation of what the name means.  (for our future selves).

File: lib/holiday-stops/src/main/scala/com/gu/holiday_stops/CurrentGuardianWeeklySubscription.scala [hunk 2]
```diff
@@ -41,11 +43,44 @@ case object ChargeIsQuarterlyOrAnnual extends CurrentGuardianWeeklyRatePlanCondi
...
+    }.getOrElse(false)
+  }
+}
+
+case object RatePlanHasNForNGuardianWeeklyIntroPlan extends CurrentGuardianWeeklyRatePlanCondition {
+  def apply(ratePlans: List[RatePlan], guardianWeeklyNForNProductRatePlanIds: List[String]): Boolean =
+    ratePlans.exists(ratePlan => guardianWeeklyNForNProductRatePlanIds.contains(ratePlan.productRatePlanId))
+}
+
+case object GuradianWeeklyNForNHasBeenInvoiced extends CurrentGuardianWeeklyRatePlanCondition {
```
Comment (kelvin-chappell) [inline lib/holiday-stops/src/main/scala/com/gu/holiday_stops/CurrentGuardianWeeklySubscription.scala]:
```suggestion
case object GuardianWeeklyNForNHasBeenInvoiced extends CurrentGuardianWeeklyRatePlanCondition {
```

File: lib/holiday-stops/src/main/scala/com/gu/holiday_stops/CurrentGuardianWeeklySubscription.scala [hunk 3]
```diff
@@ -108,18 +175,69 @@ object CurrentGuardianWeeklySubscription {
...
-          ratePlanId = currentGuardianWeeklyRatePlan.id,
-          productRatePlanId = currentGuardianWeeklyRatePlan.productRatePlanId
+          invoicedPeriod = predictedInvoicePeriod,
+          ratePlanId = currentGuardianWeeklyWithoutInvoice.id,
+          productRatePlanId = currentGuardianWeeklyWithoutInvoice.productRatePlanId
         )
-      }.toRight(ZuoraHolidayWriteError(s"Subscription does not have a current Guardian Weekly rate plan: ${subscription}"))
+      }
+
+    (maybeRegularGw orElse maybeGwWithIntro)
```
Comment (kelvin-chappell) [inline lib/holiday-stops/src/main/scala/com/gu/holiday_stops/CurrentGuardianWeeklySubscription.scala]: *** This looks like the key line in the whole thing. Just to orientate myself.

**Comment by twrichards** [comment]
Does this just handle the scenario where the issue date being processed occurs during the main rate plan, but processing runs before that rate plan has been invoiced for the first time OR does it also handle/honour when the issue date occurs during the intro plan and thus chooses that rate plan for the amount calculation (i.e. credit them a £1)?

**Comment by mario-galic** [comment]
Unfortunately supporting holidays within intro period represents more significant change to the rest of codebase. I will address this in separate PR. For now I have placed a guard against such scenario:

```
      _ <- if (currentGuardianWeeklySubscription.introNforNMode && stop.stoppedPublicationDate.isBefore(currentGuardianWeeklySubscription.invoicedPeriod.startDateIncluding)) Left(ZuoraHolidayWriteError("Holiday within intro N-for-N plan is not currently supported.")) else Right(())

```

**Pull request merged**